src/java.base/share/classes/java/text/NumberFormat.java

```diff
@@ -347,7 +347,7 @@ public final String format(double number) {
         if (result != null)
             return result;
 
-        return format(number, new StringBuffer(),
```

Comment:
Instead of just calling the package-private method that cannot be implemented by user subclasses, we can probably add a package-private method like `boolean isInternalSubclass` to decide if we should call the Buffer or Builder version.